dataplane/api-server/src/netutils.rs

```diff
@@ -5,59 +5,73 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
-///
-/// TODO: replace this https://github.com/Kong/blixt/issues/49
-pub fn if_name_for_routing_ip(ip_addr: Ipv4Addr) -> Result<String, Error> {
-    // run the linux command "ip route" to get the device responsible for
-    // routing the given IP address.
-    let ip = ip_addr.to_string();
-    let mut cmd = Command::new("ip");
-    let child = cmd
-        .arg("route")
-        .arg("get")
-        .arg("to")
-        .arg(&ip)
-        .stdout(Stdio::piped())
-        .spawn()?;
-
-    // grab the output from the "ip route" command
-    let output = child.wait_with_output()?;
-    let stdout = from_utf8(output.stdout.as_slice())?;
-
-    // construct a regex to match the output
-    let mut regex_str = ip.clone();
-    regex_str.push_str(r" (via [0-9]+\.[0-9]+\.[0-9]+\.[0-9]+ )?dev ([a-zA-Z0-9]+)\s+");
-    let re = Regex::new(&regex_str)?;
-
-    // match on the output to find the network device responsible for routing
-    // the provided IP address.
-    let match_err = format!("no device found to route {}", ip);
-    let device = re
-        .captures(stdout)
-        .ok_or_else(|| Error::msg(match_err.clone()))?
-        .iter()
-        .last()
-        .ok_or_else(|| Error::msg(match_err.clone()))?
-        .ok_or_else(|| Error::msg(match_err))?
-        .as_str()
-        .to_owned();
-
-    Ok(device)
+/// Returns an network interface index for a Ipv4 address (like the command `ip route get to $IP`)
+pub fn if_index_for_routing_ip(ip_addr: Ipv4Addr) -> Result<u32, Error> {
+    let mut socket = Socket::new(NETLINK_ROUTE)?;
+    let _port_number = socket.bind_auto()?.port_number();
+    socket.connect(&SocketAddr::new(0, 0))?;
+
+    let mut nl_hdr = NetlinkHeader::default();
+    nl_hdr.flags = NLM_F_REQUEST | NLM_F_DUMP_FILTERED;
+
+    // construct RouteMessage
+    let route_header = RouteHeader {
+        address_family: AddressFamily::Inet,
+        flags: RouteFlags::LookupTable,
+        destination_prefix_length: 32,
+        table: RouteHeader::RT_TABLE_MAIN,
+        ..Default::default()
+    };
+    let route_attribute = RouteAttribute::Destination(RouteAddress::Inet(ip_addr));
+    let mut route_message = RouteMessage::default();
+    route_message.attributes = vec![route_attribute];
+    route_message.header = route_header;
+
+    let no_ifindex_err = format!("no ifindex found to route {}", ip_addr);
+    let con_packet_err = "construct packet failed".to_string();
+
+    let mut packet = NetlinkMessage::new(
+        nl_hdr,
+        NetlinkPayload::from(RouteNetlinkMessage::GetRoute(route_message)),
+    );
+    packet.finalize();
+    let mut buf = vec![0; packet.header.length as usize];
+    // check packet
+    if buf.len() != packet.buffer_len() {
+        return Err(Error::msg(con_packet_err));
+    }
+    packet.serialize(&mut buf[..]);
+    socket.send(&buf[..], 0)?;
+
+    let mut receive_buffer = vec![0; 4096];
+    socket.recv(&mut &mut receive_buffer[..], 0)?;
+
+    let bytes = &receive_buffer[..];
```

Comment:
tiny thing, but since we only use this once from here on, perhaps we don't need the extra variable and can just hand the `receive_buffer` to `deserialize`?